libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,509 @@
+# TickFS Technical Details
+
+## How TickFS works
+
+Unlike a regular File System (FS) TickFS is only designed to store Key/Value (KV)
+pairs in flash. It does not support writing actual files, directories or other
+complex objects. Although a traditional file system layer could be added on top
+to add such features.
+
+TickFS allows writing new key/value pairs (by appending them) and removing
+old key/value pairs.
+
+Similar to Yaffs1 TickFS uses a log structure and will circle over the flash
+data. This means that the file system is inherently wear leveling as we don't
+regularly write to the same flash region.
+
+## Storage Format
+
+TickFS flash storage is formatted as a sorting of regions with each region
+containing TickFS objects.
+
+### TickFS Regions
+
+A TickFS region is the smallest set of contiguous flash storage space that can
```

Comment:
Good questions.

Right now it can not be larger. The main reason is to avoid having to do multiple reads once we have found a key. Also writing over flash bounds will be a little more tricky. I think this is something that could be added in the future. It's not a fundamental limitation, just an implementation limitation.

There is a KV header entry. Right now it's empty. I will look at saving the region size in there.

Reply:
You mention multi-page writes: what are the durability and atomicity guarantees of flash writes/erases? What happens if you lose power in the middle of a write/erase? If you do not have atomicity guarantees here then you have to design the system differently to provide ACID semantics.

Reply:
Aren't you always going to have that problem though? If we loss power when some data has been written, but not all the data then we are going to loose that key. Is there much we can do about that?

Reply:
You do not always have that problem: https://en.wikipedia.org/wiki/Write-ahead_logging

This is how all modern file systems and databases are built. Filesystems usually rely on block atomicity, which storage units provide (i.e., have enough capacitance to complete an operation once started). But they use this block atomicity to build higher-level atomicity. Note that in RAM-scarce settings, WAL does require two writes.

It's OK if TicKV doesn't provide full ACID. But you should be very clear about what guarantees it does provide.

Reply:
Ah, yep. WAL doesn't really work on flash though, at least not without wear levelling issues.

Ok, I think I have described that. We are dependent on the flash implementation for the operations. Let me know if you think I should change the README.